File: P3/Civitas/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.Respuestas;
import civitas.TituloPropiedad;

//Se cambia visibilidad de paquete a public para usarlo en el main de prueba
public class VistaTextual {

    CivitasJuego juegoModel;
    private int iGestion = -1;
    private int iPropiedad = -1;
    private static String separador = "=====================";

    private Scanner in;

    //Se cambia visibilidad de paquete a public para usarlo en el main de prueba
    public VistaTextual() {
        in = new Scanner(System.in);
    }

    void mostrarEstado(String estado) {
        System.out.println(estado);
    }

    void pausa() {
        System.out.print("Pulsa una tecla");
        in.nextLine();
    }

    int leeEntero(int max, String msg1, String msg2) {
        Boolean ok;
        String cadena;
        int numero = -1;
        do {
            System.out.print(msg1);
            cadena = in.nextLine();
            try {
                numero = Integer.parseInt(cadena);
                ok = true;
            } catch (NumberFormatException e) { // No se ha introducido un entero
                System.out.println(msg2);
                ok = false;
            }
            if (ok && (numero < 0 || numero >= max)) {
                System.out.println(msg2);
                ok = false;
            }
        } while (!ok);

        return numero;
    }

    int menu(String titulo, ArrayList<String> lista) {
        String tab = "  ";
        int opcion;
        System.out.println(titulo);
        for (int i = 0; i < lista.size(); i++) {
            System.out.println(tab + i + "-" + lista.get(i));
        }

        opcion = leeEntero(lista.size(),
                "\n" + tab + "Elige una opción: ",
                tab + "Valor erróneo");
        return opcion;
    }

    SalidasCarcel salirCarcel() {
        int opcion = menu("Elige la forma para intentar salir de la carcel",
                new ArrayList<>(Arrays.asList("Pagando", "Tirando el dado")));
        return (SalidasCarcel.values()[opcion]);
    }

    Respuestas comprar() {
        int opcion = menu("¿Desea comprar la calle?",
                new ArrayList<>(Arrays.asList("NO", "SI")));
        return (Respuestas.values()[opcion]);
    }

    void gestionar() {
        iGestion = menu("¿Qué número de gestión inmobiliaria desea elegir?",
                new ArrayList<>(Arrays.asList("Vender", "Hipotecar", "Cancelar hipoteca",
                        "Construir casa", "Construir hotel", "Terminar")));
        ArrayList<String> propiedades = new ArrayList<>();
        for (int i = 0; i < juegoModel.getJugadorActual().getPropiedades().size(); i++) {
            propiedades.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());
        }
        if (iGestion != 5) {
            iPropiedad = menu("Indique la propiedad a la que desea aplicar la gestión: ", propiedades);
        }

    }

    public int getGestion() {
        return iGestion;
    }

    public int getPropiedad() {
        return iPropiedad;
    }

    void mostrarSiguienteOperacion(OperacionesJuego operacion) {
        System.out.println("Siguiente operación que se va a realizar");
        System.out.println(operacion);
    }

    void mostrarEventos() {
        while (Diario.getInstance().eventosPendientes()) {
            System.out.println(Diario.getInstance().leerEvento());
        }
    }

    public void setCivitasJuego(CivitasJuego civitas) {
        juegoModel = civitas;
        this.actualizarVista();
    }

    public void actualizarVista() {
        System.out.println("Jugador actual: ");
        System.out.println(juegoModel.infoJugadorTexto());
        System.out.println("Posicion actual: ");
        System.out.println(juegoModel.getCasillaActual());
    }

}

File: P3/civitasruby/lib/juego_texto/vista_textual.rb
#encoding:utf-8
require_relative '../civitas/operaciones_juego'
require 'io/console'
require_relative '../civitas/civitas_juego'
require_relative '../civitas/salidas_carcel'
require_relative '../civitas/respuestas'
require_relative '../civitas/gestiones_inmobiliarias'
require_relative '../civitas/diario'
module Civitas

  class Vista_textual

    def initialize()
      @iPropiedad
      @iGestion
      @juegoModel
    end

    def mostrar_estado(estado)
      puts estado
    end


    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l.to_s
        index += 1
      }

      opcion = lee_entero(lista.length,
        "\n"+tab+"Elige una opción: ",
        tab+"Valor erróneo")
      return opcion
    end


    def comprar
      opcion=menu("¿Deseas comprar la calle?", ["No", "Si"])
      Respuestas::Lista_respuestas.at(opcion)
    end

    def gestionar
      opciones = ["Vender", "Hipotecar", "Cancelar Hipoteca", "Construir Casa", "Construir Hotel", "Terminar" ]
      elegido = menu("Elige la gestión inmobiliaria",opciones)
      @iGestion=elegido
      if elegido != 5
        propiedad = menu("Elige la propiedad que quieres gestionar",@juegoModel.get_jugador_actual.get_propiedades )
        @iPropiedad = propiedad
      end
    end

    def get_gestion
      @iGestion
    end

    def get_propiedad
      @iPropiedad
    end

    def mostrar_siguiente_operacion(operacion)
      print "Siguiente operacion #{operacion}\n"
    end

    def mostrar_eventos
      while Diario.instance.eventos_pendientes
        puts Diario.instance.leer_evento
      end
    end

    def set_civitas_juego(civitas)
      @juegoModel=civitas
      actualizar_vista
    end

    def actualizar_vista
      jugador = @juegoModel.get_jugador_actual
      casilla = @juegoModel.get_casilla_actual
      puts "Jugador actual"
      puts jugador.to_s
      puts "Casilla actual"
      puts casilla.to_s
    end

    def salir_carcel
      opcion=menu("¿Cómo quieres salir de la cárcel",["Pagando","Tirando"])
      Salidas_carcel::Lista_salidas_carcel.at(opcion)
    end

  end

end
